concert/ext/tangoservers/porqa.py

```diff
@@ -0,0 +1,141 @@
+"""
+porqa.py
+---------
+Implements a device server to execute post-reconstruction quality assurance routines.
+"""
+import logging
+from typing import AsyncIterator, List
+import numpy as np
+import matplotlib.pyplot as plt
+try:
+    import cupy as xp
+    import cupy.fft as xft
+except ModuleNotFoundError:
+    print("cupy not available, defaulting to numpy")
+    import numpy as xp
+    import numpy.fft as xft
+from tango import DebugIt, DevState
+from tango.server import attribute, AttrWriteType, command
+from tqdm import tqdm
+from concert.ext.tangoservers.base import TangoRemoteProcessing
+from concert.typing import ArrayLike
+
+
+class TangoPostReconstructionQA(TangoRemoteProcessing):
+    """Tango device for post-reconstruction quality assurance on a remote server"""
+
+    freq_bins = attribute(
+        label="Number of frequency bins",
+        dtype=int,
+        access=AttrWriteType.WRITE,
+        fget="get_freq_bins",
+        fset="set_freq_bins",
+        doc="number of frequency bins for Fourier Shell Correlation (FSC) calculation"
+    )
+
+    plot = attribute(
+        label="Plot",
+        dtype=bool,
+        access=AttrWriteType.WRITE,
+        fget="get_plot",
+        fset="set_plot",
+        doc="Plot of the Fourier Shell Correlation (FSC) in different directions"
+    )
+
+    async def init_device(self) -> None:
+        """
+        Initializes the remote walker tango device. Sets the remote server's
+        home directory as root as well as the current directory.
+        """
+        logging.getLogger('matplotlib.font_manager').disabled = True
+        await super().init_device()
+        self.set_state(DevState.STANDBY)
+        self.info_stream(
+            "%s in state: %s", self.__class__.__name__, self.get_state())
+        
+    def get_freq_bins(self) -> int:
+        return self._freq_bins
+    
+    def set_freq_bins(self, fb: int) -> None:
+        self._freq_bins = fb
+        self.info_stream("%s: set frequency bins to %d", self.__class__.__name__, fb)
+
+    def get_plot(self) -> bool:
+        return self._plot
+    
+    def set_plot(self, plot: bool) -> None:
+        self._plot = plot
+        self.info_stream("%s: set plot to %s", self.__class__.__name__, plot)
+
+    @DebugIt(show_args=True)
+    @command()
+    async def exec_qa(self) -> None:
+        await self._process_stream(self._exec_fsc(self._receiver.subscribe()))
+
+    def _compute_directional_fsc(self, fft1: ArrayLike, fft2: ArrayLike,
+                                 direction: str = 'xy') -> ArrayLike:
+        self.info_stream("%s: computing SFSC for direction %s", self.__class__.__name__, direction)
+        z, y, x = np.indices(fft1.shape)
+        center: ArrayLike = np.array(fft2.shape) // 2
+        rz, ry, rx = z - center[0], y - center[1], x - center[2]
+        # Directional distance
+        if direction == 'xy':
+            r: float = np.sqrt(rx**2 + ry**2)
```

Comment:
This makes a cylinder, if you want to exclude "z" for example, you should calculate in _plane_ z=0, otherwise you are mixing things. It will also speed up computation because you won't need 3D volumes, only 2D slices. Let's discuss this when you are back, just put a pin in this for now.